### ISSUE TITLE
Crash on printf '\e[81C '

### ISSUE THREAD

**Issue by andersk** [jwilm/alacritty#133]
When running `printf '\e[81C\eH'`:

```
thread 'pty reader' panicked at 'index out of bounds: the len is 80 but the index is 81', /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libcollections/vec.rs:1371
stack backtrace:
   1:     0x5588ad085d4c - std::sys::imp::backtrace::tracing::imp::write::h9c41d2f69e5caabf
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:42
   2:     0x5588ad08b23e - std::panicking::default_hook::{{closure}}::h1f61f3c769fffe7a
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:351
   3:     0x5588ad08ae44 - std::panicking::default_hook::hd5bda4e453dfb4be
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:367
   4:     0x5588ad08b6e7 - std::panicking::rust_panic_with_hook::hffbc74969c7b5d87
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:555
   5:     0x5588ad08b574 - std::panicking::begin_panic::h3bbdbb2a23440035
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:517
   6:     0x5588ad08b499 - std::panicking::begin_panic_fmt::h34f5b320b0f94559
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:501
   7:     0x5588ad08b427 - rust_begin_unwind
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:477
   8:     0x5588ad0b75bd - core::panicking::panic_fmt::h1016b85b51d1931f
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libcore/panicking.rs:69
   9:     0x5588ad0b7568 - core::panicking::panic_bounds_check::h3b875b6ea9a674f6
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libcore/panicking.rs:56
  10:     0x5588accec545 - <collections::vec::Vec<T> as core::ops::IndexMut<usize>>::index_mut::h69b598477d3e900a
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libcollections/vec.rs:1371
  11:     0x5588acc5fa44 - <alacritty::grid::Row<T> as core::ops::IndexMut<alacritty::index::Column>>::index_mut::h9adf91b5b10bafe3
                        at /home/anders/rust/alacritty/src/grid.rs:311
  12:     0x5588acc64280 - <alacritty::grid::Grid<T> as core::ops::IndexMut<&'point alacritty::index::Point>>::index_mut::h2a8363d21414e0ce
                        at /home/anders/rust/alacritty/src/grid.rs:225
  13:     0x5588accdaef7 - <alacritty::term::Term as alacritty::ansi::Handler>::input::h1bacbff20d66275a
                        at /home/anders/rust/alacritty/src/term/mod.rs:638
  14:     0x5588accf6d6c - <alacritty::ansi::Performer<'a, H, W> as vte::Perform>::print::h2f59924c294298c6
                        at /home/anders/rust/alacritty/src/ansi.rs:438
  15:     0x5588acc80616 - vte::Parser::perform_action::h82162452c9ac296a
                        at /home/anders/.cargo/registry/src/github.com-1ecc6299db9ec823/vte-0.1.2/src/lib.rs:189
  16:     0x5588acc80c4e - vte::Parser::perform_state_change::ha318e6e99fbda378
                        at /home/anders/.cargo/registry/src/github.com-1ecc6299db9ec823/vte-0.1.2/src/lib.rs:167
  17:     0x5588acc80fc4 - vte::Parser::advance::h66c6bd0670baf377
                        at /home/anders/.cargo/registry/src/github.com-1ecc6299db9ec823/vte-0.1.2/src/lib.rs:137
  18:     0x5588accfbfac - alacritty::ansi::Processor::advance::h92e3d5ba20d53d25
                        at /home/anders/rust/alacritty/src/ansi.rs:85
  19:     0x5588accc6986 - <alacritty::event_loop::EventLoop<Io>>::pty_read::h18a7cc99c622d42f
                        at /home/anders/rust/alacritty/src/event_loop.rs:211
  20:     0x5588accc6082 - <alacritty::event_loop::EventLoop<Io>>::spawn::{{closure}}::h069a2120b2d00625
                        at /home/anders/rust/alacritty/src/event_loop.rs:308
  21:     0x5588accf8024 - <std::panic::AssertUnwindSafe<F> as core::ops::FnOnce<()>>::call_once::h6032dcb0cf344702
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panic.rs:296
  22:     0x5588acc8048f - std::panicking::try::do_call::hc745aec8a9cedc1b
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:460
  23:     0x5588ad09265a - __rust_maybe_catch_panic
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libpanic_unwind/lib.rs:98
  24:     0x5588acc7fb03 - std::panicking::try::h697849a0f4444602
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:436
  25:     0x5588acc7dbb4 - std::panic::catch_unwind::hb5b74bacab7a9eb7
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panic.rs:361
  26:     0x5588acc7f45c - std::thread::Builder::spawn::{{closure}}::hc2b962b0b41adef6
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/thread/mod.rs:357
  27:     0x5588accbf2e9 - <F as alloc::boxed::FnBox<A>>::call_box::hab7d1897638a2589
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/liballoc/boxed.rs:605
  28:     0x5588ad08a534 - std::sys::imp::thread::Thread::new::thread_start::h76badbf9b0ecaf58
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/liballoc/boxed.rs:615
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/sys_common/thread.rs:21
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/sys/unix/thread.rs:84
  29:     0x7f76efc8c6c9 - start_thread
  30:     0x7f76ef7af0ae - __clone
  31:                0x0 - <unknown>
child finished with error '1'
```

**Comment by gmorenz** [comment]
Can be shortened to `printf '\e[81C`

Specific issue is that [this line](https://github.com/jwilm/alacritty/blob/4fad82bc267141532460d6fc15b1ee6c1acd9589/src/ansi.rs#L528) blindly accepts a column to go to without checking that it's a valid column. As does the `move_forward` function.

**Comment by andersk** [comment]
The crash doesn’t happen on just `'\e[81C'`, but it happens if something is printed after that.

I don’t think `move_forward` is the right place for the check. That would have a TOCTTOU bug if the window was resized between the move and the following print.

**Comment by gmorenz** [comment]
> The crash doesn’t happen on just '\e[81C', but it happens if something is printed after that.

Right, it's just that my (any really) PS1 prompt will trigger it after that.

> I don’t think move_forward is the right place for the check. That would have a TOCTTOU bug if the window was resized between the move and the following print.

I'm not really familiar with the architecture of this program, but at first glance it seems like the appropriate place to me. 

Resize has to handle fixing cursor location as well (already does it seems), since you might be shrinking the window to before where the cursor is. So resizing the window after doing `printf \e[81C` shouldn't cause any problems, in theory at least. 

Meanwhile the escape sequence appears to be specified to stop at the right edge of the screen (at least it is [here](http://www.inwap.com/pdp10/ansicode.txt), is there some official documentation on these?) so if we don't do it there, or shortly after there, something like "move forwards 100000, move back 10" is going to be broken.

**Comment by andersk** [comment]
Seems fixed by #164.
